app/src/main/kotlin/io/homeassistant/companion/android/widgets/BaseGlanceEntityWidgetReceiver.kt

```diff
@@ -0,0 +1,284 @@
+package io.homeassistant.companion.android.widgets
+
+import android.appwidget.AppWidgetManager
+import android.content.Context
+import android.content.Intent
+import android.content.IntentFilter
+import androidx.annotation.VisibleForTesting
+import androidx.core.content.ContextCompat
+import androidx.glance.appwidget.GlanceAppWidgetManager
+import androidx.glance.appwidget.GlanceAppWidgetReceiver
+import androidx.glance.appwidget.updateAll
+import io.homeassistant.companion.android.common.data.integration.Entity
+import io.homeassistant.companion.android.common.data.servers.ServerManager
+import io.homeassistant.companion.android.database.widget.WidgetDao
+import javax.inject.Inject
+import kotlinx.coroutines.CoroutineExceptionHandler
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.cancel
+import kotlinx.coroutines.isActive
+import kotlinx.coroutines.launch
+import timber.log.Timber
+
+private fun newCoroutineScopeProvider(): () -> CoroutineScope {
+    val exceptionHandler = CoroutineExceptionHandler { _, throwable -> Timber.e(throwable, "Unhandled error in widget scope") }
+    // Use SupervisorJob to avoid cancelling all jobs when one fails since the scope is used across all the widgets
+    return { CoroutineScope(Dispatchers.Main + SupervisorJob() + exceptionHandler) }
+}
+
+data class EntitiesPerServer(val serverId: Int, val entityIds: List<String>)
+
+/**
+ * Base class for Glance widgets that update based on entity state changes.
+ *
+ * This class provides the foundational functionality for managing widget updates, handling lifecycle events,
+ * and observing entity state changes. It is designed to be extended by specific widget implementations.
+ *
+ * ### Key Features:
+ * - **Entity state observation**: Watches for updates to specific entities and triggers widget updates accordingly.
+ * - **Lifecycle management**: Handles widget lifecycle action [Intent.ACTION_SCREEN_ON], [Intent.ACTION_SCREEN_OFF] and [AppWidgetManager.ACTION_APPWIDGET_UPDATE].
+ * - **Real-time updates**: Supports real-time updates for widgets when the user grants the necessary permissions.
```

Comment:
What is "the necessary permission"? Also applies to lines 47-48. Is that documentation needed here or best placed elsewhere?

Reply:
@dshokouhi I don't know if this has a permission name do you know?

Reply:
I didn't find anything on the doc speaking about this page 
![image](https://github.com/user-attachments/assets/2d39e412-162a-437b-bae6-a3f4c9c91d12)
it's anoying that this page control widget refresh state... I think it's kinda linked to https://developer.android.com/develop/ui/views/notifications/conversations#conversation-widgets since they need to stay up to date.

Reply:
this is hte intent: https://developer.android.com/reference/android/provider/Settings#ACTION_NOTIFICATION_LISTENER_SETTINGS

and this is hte permission: https://developer.android.com/reference/android/Manifest.permission#BIND_NOTIFICATION_LISTENER_SERVICE

Reply:
That is not a widget-related permission though, it makes the system run a service for the app all the time. Presumably you'd see the same behavior for the widget if you enabled the websocket foreground service?

It may be that recent Android changes now mean that you need to have some part of the app active to keep the widget's coroutinescope (and thus subscription) working, whereas previously it worked without, but I would not point to notification access here. That is just misleading IMO.

Reply:
I just tested and even with the foreground service up and running. If you don't gives this "real time" permission the widget doesn't refresh after it exits the composition state... 

I know this permission is for notification, but I think Google is using it for the conversation widget to stay in sync all the time. It's only an assumption but it kinda sounds ok with how they've documented it.

This is kind of a hack of the API to keep a widget up to date. We should document that in the companion docs. I wrote it here for future contributors to understand that by default it won't stay up to date.

The reality is that the default behavior is that the widget will enter composition every 30mn for 1mn not more on new version of Android, even if an event occurs it won't do anything.

Reply:
https://github.com/home-assistant/companion.home-assistant/pull/1203 it's a proposal.